### PR TITLE
Allow keyboard shortcuts to work on native devices

### PR DESCRIPTION
**Pull request by azimgd**
### Fixed Issues

$ [6883](https://github.com/Expensify/App/issues/6883)   
PROPOSAL: [6883](https://github.com/Expensify/App/issues/6883)   

### Tests

- [ ] Verify that no errors appear in the JS console

### QA Steps

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [ ] I linked the correct issue in the `### Fixed Issues` section above
- [ ] I wrote clear testing steps that cover the changes made in this PR
    - [ ] I added steps for local testing in the `Tests` section
    - [ ] I added steps for the expected offline behavior in the `Offline steps` section
    - [ ] I added steps for Staging and/or Production testing in the `QA steps` section
    - [ ] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [ ] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [ ] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I ran the tests on **all platforms** & verified they passed on:
    - [ ] Android / native
    - [ ] Android / Chrome
    - [ ] iOS / native
    - [ ] iOS / Safari
    - [ ] MacOS / Chrome / Safari
    - [ ] MacOS / Desktop
- [ ] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [ ] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [ ] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [ ] I verified that comments were added to code that is not self explanatory
    - [ ] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [ ] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [ ] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [ ] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [ ] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [ ] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [ ] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [ ] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [ ] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [ ] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [ ] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [ ] I verified that if a function's arguments changed that all usages have also been updated correctly
- [ ] If a new component is created I verified that:
    - [ ] A similar component doesn't exist in the codebase
    - [ ] All props are defined accurately and each prop has a `/** comment above it */`
    - [ ] The file is named correctly
    - [ ] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [ ] The only data being stored in the state is data necessary for rendering and nothing else
    - [ ] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [ ] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [ ] All JSX used for rendering exists in the render method
    - [ ] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [ ] If any new file was added I verified that:
    - [ ] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [ ] If a new CSS style is added I verified that:
    - [ ] A similar style doesn't already exist
    - [ ] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [ ] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [ ] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [ ] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [ ] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

</details>

<details>
<summary>Mobile Web - Chrome</summary>

</details>

<details>
<summary>Mobile Web - Safari</summary>

</details>

<details>
<summary>Desktop</summary>

</details>

<details>
<summary>iOS</summary>

</details>

<details>
<summary>Android</summary>

</details>

### CODE REVIEW

**Comment by azimgd** [comment]
waiting for the package to be published.

**Comment by luacmartins** [comment]
Sorry, I didn't get to this one today. I'll get back to it tomorrow!

**Comment by luacmartins** [comment]
@azimgd left a [comment](https://github.com/Expensify/App/issues/6883#issuecomment-1381108076) about publishing the package to npm

**Comment by azimgd** [comment]
@luacmartins wrapping up my PR and will un-publish the package tomorrow.

**Comment by luacmartins** [comment]
Ok, let me know when it's done and I'll work on publishing it from our end. Thanks!

**Comment by luacmartins** [comment]
@azimgd did you get a chance to unpublish the package?

**Comment by azimgd** [comment]
yes. 👍

**Comment by luacmartins** [comment]
Ok, I tried publishing but it seems like there's a 24h wait period. I'll try again tomorrow!

>403 Forbidden - PUT https://registry.npmjs.org/react-native-key-command - react-native-key-command cannot be republished until 24 hours have passed.

**Comment by luacmartins** [comment]
Package [published](https://www.npmjs.com/package/react-native-key-command)!

**Comment by luacmartins** [comment]
@azimgd what's the latest update on this one?

**Comment by azimgd** [comment]
Android commands:
CTRL+K: `adb shell "input keyevent 113 && input keyevent 39"`
CTRL+SHIFT+K: `adb shell "input keyevent 113 && input keyevent 39 && input keyevent 59"`
CTRL+I: `adb shell "input keyevent 113 && input keyevent 37"`
ESC: `adb shell "input keyevent 111"`

**Comment by azimgd** [comment]
Let's start with the code review when you get a chance. I would like to know if you have any comments on the approach.

Overall just couple of things are missing there:
- Using priority argument
- Debug cmd + enter shortcut

**Comment by azimgd** [comment]
@parasharrajat merged

**Review by parasharrajat — CHANGES_REQUESTED** [review]

File: ios/Podfile.lock [hunk 1]
```diff
@@ -181,25 +181,25 @@ PODS:
...
     - GoogleUtilities/Logger
     - "GoogleUtilities/NSData+zlib"
     - GoogleUtilities/Reachability
-  - "GoogleUtilities/NSData+zlib (7.10.0)"
-  - GoogleUtilities/Reachability (7.10.0):
+  - "GoogleUtilities/NSData+zlib (7.11.0)"
+  - GoogleUtilities/Reachability (7.11.0):
     - GoogleUtilities/Logger
-  - GoogleUtilities/UserDefaults (7.10.0):
+  - GoogleUtilities/UserDefaults (7.11.0):
```
[the remark this answers is not in the document]
Comment (azimgd) [reply ios/Podfile.lock]: default pod update. rolledback.

File: ios/Podfile.lock [hunk 2]
```diff
@@ -957,7 +962,7 @@ SPEC CHECKSUMS:
   glog: 04b94705f318337d7ead9e6d17c019bd9b1f6b1b
   GoogleAppMeasurement: 5ba1164e3c844ba84272555e916d0a6d3d977e91
   GoogleDataTransport: 1c8145da7117bd68bbbed00cf304edb6a24de00f
-  GoogleUtilities: bad72cb363809015b1f7f19beb1f1cd23c589f95
+  GoogleUtilities: c2bdc4cf2ce786c4d2e6b3bcfd599a25ca78f06f
```
[the remark this answers is not in the document]
Comment (azimgd) [reply ios/Podfile.lock]: done

File: src/libs/KeyboardShortcut/index.js [hunk 3]
```diff
@@ -135,6 +56,40 @@ function getPlatformEquivalentForKeys(keys) {
...
+const getLibraryAdjustedModifiers = (modifiers) => {
+    if (_.isEqual(modifiers, ['CTRL', 'SHIFT'])) {
+        return KeyCommand.constants.keyModifierShiftCommand;
+    }
+
+    if (_.isEqual(modifiers, ['CTRL'])) {
+        return KeyCommand.constants.keyModifierCommand;
+    }
+
+    return undefined;
```
[the remark this answers is not in the document]
Comment (azimgd) [reply src/libs/KeyboardShortcut/index.js:68]: removed

File: src/libs/KeyboardShortcut/index.js [hunk 4]
```diff
@@ -135,6 +56,40 @@ function getPlatformEquivalentForKeys(keys) {
...
+
+    if (_.isEqual(key, 'enter')) {
+        return KeyCommand.constants.keyInputEnter;
+    }
+
+    if (_.isEqual(key, 'arrowup')) {
+        return KeyCommand.constants.keyInputUpArrow;
+    }
+
+    if (_.isEqual(key, 'arrowdown')) {
```
[the remark this answers is not in the document]
Comment (azimgd) [reply src/libs/KeyboardShortcut/index.js]: updated

**Review by parasharrajat — CHANGES_REQUESTED** [review]
BUG: The web shortcut `CTRL+ I` does not work on firefox and chrome. It is working on staging. 

- Pressing `CTRL+ I` opens page info dialog on this branch. 


--- 

Same for `CTRL + K` & `CTRL + SHIFT + K`.

**Comment by azimgd** [comment]
Mac ?

**Comment by azimgd** [comment]
I'm well aware of the fact that it should work on any platform. I asked you which platform are you testing on.

**Comment by azimgd** [comment]
updated.

**Comment by azimgd** [comment]
@parasharrajat any more feedback?

**Comment by azimgd** [comment]
Have you had a chance to test ?

**Review by parasharrajat — CHANGES_REQUESTED** [review]

File: src/styles/styles.js [hunk 5]
```diff
@@ -2658,7 +2658,6 @@ const styles = {
 
     keyboardShortcutModalContainer: {
         maxHeight: '100%',
-        flex: '0 0 auto',
```
[the remark this answers is not in the document]
Comment (azimgd) [reply src/styles/styles.js:2661]: RN doesn't support '0 0 auto'.

File: src/components/KeyboardShortcutsModal.js [hunk 6]
```diff
@@ -83,7 +83,7 @@ class KeyboardShortcutsModal extends React.Component {
                 onClose={KeyboardShortcutsActions.hideKeyboardShortcutModal}
             >
                 <HeaderWithCloseButton title={this.props.translate('keyboardShortcutModal.title')} onCloseButtonPress={KeyboardShortcutsActions.hideKeyboardShortcutModal} />
-                <View style={[styles.p5, styles.pt0]}>
+                <ScrollView style={[styles.p5, styles.pt0]}>
```
[the remark this answers is not in the document]
Comment (azimgd) [reply src/components/KeyboardShortcutsModal.js:86]: This is needed on mobile.

File: src/libs/KeyboardShortcut/index.js [hunk 7]
```diff
@@ -135,6 +56,46 @@ function getPlatformEquivalentForKeys(keys) {
...
+
+    if (_.isEqual(modifiers, ['CTRL', 'SHIFT'])) {
+        if (_.includes([CONST.OS.MAC_OS, CONST.OS.IOS], operatingSystem)) {
+            return KeyCommand.constants.keyModifierShiftCommand;
+        }
+        return KeyCommand.constants.keyModifierControlShift;
+    }
+
+    if (_.isEqual(modifiers, ['CTRL'])) {
+        if (_.includes([CONST.OS.MAC_OS, CONST.OS.IOS], operatingSystem)) {
```
[the remark this answers is not in the document]
Comment (azimgd) [reply src/libs/KeyboardShortcut/index.js:70]: I thought of this as well, but it looked like an abstraction over abstraction for me. My point is that code reads just fine here and the only changing part between OS is a combination of `cmd` key.

File: src/styles/styles.js (hunk 5 above)
[the remark this answers is not in the document]
Comment (azimgd) [reply src/styles/styles.js:2661]: fixed by removing flexGrow on `CONST.MODAL.MODAL_TYPE.CENTERED_UNSWIPEABLE`.

That is being used at:
- KeyboardShortcutsModal
- AttachmentModal

Tested against production, results look coorect.

File: src/libs/KeyboardShortcut/index.js (hunk 7 above)
[the remark this answers is not in the document]
Comment (azimgd) [reply src/libs/KeyboardShortcut/index.js]: fixed

**Review by parasharrajat — CHANGES_REQUESTED** [review]
What are the future plans for this PR? I see that `priority` is not yet implemented. Why is this still in draft mode? 

Can you please add details of the logical changes in the PR description?

**Comment by azimgd** [comment]
Closing this in favor of https://github.com/Expensify/App/pull/14767

**Pull request closed without merging**